### PR TITLE
os/bluestore: tolerate zero length for allocators' init_[add/rm]_free()

### PR DESCRIPTION
**Pull request by ifed01**
Fixes: https://tracker.ceph.com/issues/50555
Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by tchaikov — APPROVED** [review]

File: src/test/objectstore/Allocator_test.cc [hunk 1]
```diff
@@ -47,6 +47,23 @@ TEST_P(AllocTest, test_alloc_init)
...
+  int64_t capacity = 4 * 1024 * block_size;
+
+  {
+    init_alloc(capacity, block_size);
+
+    auto free = alloc->get_free();
+    alloc->init_add_free(block_size, 0);
+    ASSERT_EQ(free, alloc->get_free());
+
+    alloc->init_rm_free(block_size, 0);
```
Comment (tchaikov) [inline src/test/objectstore/Allocator_test.cc:62]: need to update `_remove_from_tree()` and `AvlAllocator::_try_remove_from_tree()` to return early if `size == 0`.

**Review by tchaikov — CHANGES_REQUESTED** [review]

File: src/test/objectstore/Allocator_test.cc (hunk 1 above)
Comment (ifed01) [reply src/test/objectstore/Allocator_test.cc:62]: @tchaikov - actually the approach is to handle zero length earlier at init_rm_free, just missed that stuff for HybridAllocator. Fixed mw, thanks!

**Review by tchaikov — APPROVED** [review]

**Review by aclamk — APPROVED** [review]
OK.

**Comment by tchaikov** [comment]
jenkins test docs

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-01_14:22:40-rados-wip-kefu-testing-2021-05-01-1113-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-04_15:46:31-rados-wip-kefu-testing-2021-05-03-1836-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-05_09:38:15-rados-wip-kefu-testing-2021-05-05-1449-distro-basic-smithi/

i still need to look into the failures.

**Pull request merged**